Add float16 support for NonMaxSuppressionV{2,3,4}

yongtang opened the pull request:
This fix tries to address the issue raised in #20199 where there was no float16 support for NonMaxSuppressionV2. As NonMaxSuppressionV2 is the earlier versions of API
and there are newer versions of NonMaxSuppression: NonMaxSuppressionV2, NonMaxSuppressionV3, NonMaxSuppressionV4, this fix exposes the float16 support to all of the above.
(Note in the master the default version used is NonMaxSuppressionV3)

This fix fixes #20199.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang committed:
Add float16 support for NonMaxSuppressionV{2,3,4}

This fix tries to address the issue raised in 20199 where
there was no float16 support for NonMaxSuppressionV2.
As NonMaxSuppressionV2 is the earlier versions of API
and there are newer versions of NonMaxSuppression:
NonMaxSuppressionV2, NonMaxSuppressionV3, NonMaxSuppressionV4,
This fix exposes the float16 support to all of the above.
(Note in the master the default version used is NonMaxSuppressionV3)

This fix fixes 20199.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang committed:
Add test cases for float16 support for non_max_suppression

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang committed:
Pylint fix

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang committed:
Update tensorflow/core/kernels/non_max_suppression_op.cc for `Experimental clang-format Check` fix.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

azaks2 commented on tensorflow/core/kernels/non_max_suppression_op.cc:
```diff
@@ -143,13 +146,13 @@ void DoNonMaxSuppressionOp(
       candidate_priority_queue(cmp);
   for (int i = 0; i < scores_data.size(); ++i) {
-    if (scores_data[i] > score_threshold) {
+    if (scores_data[i] > static_cast<T>(score_threshold)) {
```
would it be better to upcast scores_data[i] to float.

yongtang replied:
Thanks @azaks2. The PR has been updated to use upcast instead.

azaks2 reviewed (commented):
take a look at this error
tensorflow/core/ops/compat/backwards_compatibility_test.cc:36
Expected equality of these values:
  ::tensorflow::Status::OK()
    Which is: OK
  (compatibility.ValidateCompatible(env, &changed_ops, &added_ops, nullptr))
    Which is: Invalid argument: Incompatible Op change: Attr 'T' added without default; old: Op<name=NonMaxSuppressionV2; signature=boxes:float, scores:float, max_output_size:int32, iou_threshold:float -> selected_indices:int32>; new: Op<name=NonMaxSuppressionV2; signature=boxes:T, scores:T, max_output_size:int32, iou_threshold:float -> selected_indices:int32; attr=T:type,allowed=[DT_HALF, DT_FLOAT]>

yongtang committed:
Upcast to float for better conversion, based on review feedback.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang committed:
Add default type as DT_FLOAT to maintain backward-compatibility

and fix test failure in:
```
//tensorflow/core/ops/compat:backwards_compatibility_test
```

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang commented:
Thanks @azaks2 for the review. The PR has been updated and the previous test failure:
```
//tensorflow/core/ops/compat:backwards_compatibility_test
```
has also been fixed. Please take a look and let me know if there are any issues.
